plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultQueryBuilder.java

```diff
@@ -182,7 +189,9 @@ public PreparedStatement prepareStatement(
         for (int i = 0; i < parameters.size(); i++) {
             QueryParameter parameter = parameters.get(i);
             int parameterIndex = i + 1;
-            WriteFunction writeFunction = getWriteFunction(client, session, connection, parameter.getJdbcType(), parameter.getType());
+            WriteFunction writeFunction = parameter.getJdbcType()
+                    .map(jdbcType -> getWriteFunction(client, session, connection, jdbcType, parameter.getType()))
+                    .orElseGet(() -> getWriteFunction(client, session, parameter.getType()));
```

Comment:
I don't understand the case where `toColumnMapping` will be absent for some type but `toWriteMapping` will be present.

Reply:
the key here is that we don't have `jdbcType` (a `JdbcTypeHandle`).
the  `JdbcTypeHandle` comes from a remote database, so we have it for source columns. 
So we also have it for predicates directly on columns (tuple domain), but we don't have it for complex expression pushdown. (In the first iteration i was synthesizing type handles, but they were different from those coming from the remote database, causing confusion in JdbcClient implementations. The current version avoids that)

Reply:
got it, thanks for explaining how the code evolved as well.